raft/src/main/java/org/apache/kafka/raft/CandidateState.java

```diff
@@ -235,6 +240,15 @@ public int epoch() {
         return highWatermark;
     }
 
+    @Override
+    public boolean canGrantVote(int candidateId, boolean isLogUpToDate) {
+        // Still reject vote request even candidateId = localId, Although the candidate votes for
+        // itself, this vote is implicit and not "granted".
+        log.debug("Rejecting vote request from candidate {} since we are already candidate in epoch {}",
+            candidateId, epoch);
```

Comment:
As a general rule, methods like this should not log IMO. The calling method should log instead. That is, good to avoid the side effect from the "check" operation. It seems like it was like that before this PR. What was the motivation for changing it?

Reply:
@ijuma , we logged both vote result and the reject reason in `KafkaRaftClient` before this PR, but different `EpochState` have different reject reason, to reduce the cyclomatic complexity we moved the if-else out of `KafkaRaftClient`.

Reply:
I think it's a misguided goal to reduce cyclomatic complexity if it results in surprising code. Generally, predicate type methods (`isX` or `canX`) should be callable multiple types without undesired behavior like several log lines. Maybe you can address this in a subsequent PR?

Reply:
Thank you for reminding me of this principle, maybe a better way to fix this is to return `Tuple<Boolean, String>` to represent "canGrantVote" and "rejectReason", I will try to fix this later.

Reply:
You could return Optional<String> probably where defined string would be the rejection reason. This would mean renaming the methods slightly though.